File: output/go/tests/12_state_var_push_pop.go

package main

import "fmt"
import "os"

type StateVarPushPopFrameEvent struct {
    _message    string
    _parameters map[string]any
}

type StateVarPushPopFrameContext struct {
    _event  StateVarPushPopFrameEvent
    _return any
    _data   map[string]any
}

type StateVarPushPopCompartment struct {
    state            string
    stateArgs        map[string]any
    stateVars        map[string]any
    enterArgs        map[string]any
    exitArgs         map[string]any
    forwardEvent     *StateVarPushPopFrameEvent
    parentCompartment *StateVarPushPopCompartment
}

func newStateVarPushPopCompartment(state string) *StateVarPushPopCompartment {
    return &StateVarPushPopCompartment{
        state:    state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
    }
}

func (c *StateVarPushPopCompartment) copy() *StateVarPushPopCompartment {
    nc := &StateVarPushPopCompartment{
        state: c.state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
        forwardEvent:     c.forwardEvent,
        parentCompartment: c.parentCompartment,
    }
    for k, v := range c.stateArgs { nc.stateArgs[k] = v }
    for k, v := range c.stateVars { nc.stateVars[k] = v }
    for k, v := range c.enterArgs { nc.enterArgs[k] = v }
    for k, v := range c.exitArgs { nc.exitArgs[k] = v }
    return nc
}

type StateVarPushPop struct {
    _state_stack []*StateVarPushPopCompartment
    __compartment *StateVarPushPopCompartment
    __next_compartment *StateVarPushPopCompartment
    _context_stack []StateVarPushPopFrameContext
}

func NewStateVarPushPop() *StateVarPushPop {
    s := &StateVarPushPop{}
    s._state_stack = make([]*StateVarPushPopCompartment, 0)
    s._context_stack = make([]StateVarPushPopFrameContext, 0)
    s.__compartment = newStateVarPushPopCompartment("Counter")
    s.__next_compartment = nil
    __frame_event := StateVarPushPopFrameEvent{_message: "$>", _parameters: nil}
    __ctx := StateVarPushPopFrameContext{_event: __frame_event, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return s
}

func (s *StateVarPushPop) __kernel(__e *StateVarPushPopFrameEvent) {
    s.__router(__e)
    for s.__next_compartment != nil {
        next_compartment := s.__next_compartment
        s.__next_compartment = nil
        exit_event := &StateVarPushPopFrameEvent{_message: "<$", _parameters: s.__compartment.exitArgs}
        s.__router(exit_event)
        s.__compartment = next_compartment
        if s.__compartment.forwardEvent == nil {
            enter_event := &StateVarPushPopFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
            s.__router(enter_event)
        } else {
            forward_event := s.__compartment.forwardEvent
            s.__compartment.forwardEvent = nil
            if forward_event._message == "$>" {
                s.__router(forward_event)
            } else {
                enter_event := &StateVarPushPopFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
                s.__router(enter_event)
                s.__router(forward_event)
            }
        }
    }
}

func (s *StateVarPushPop) __router(__e *StateVarPushPopFrameEvent) {
    switch s.__compartment.state {
    case "Counter":
        s._state_Counter(__e)
    case "Other":
        s._state_Other(__e)
    }
}

func (s *StateVarPushPop) __transition(next *StateVarPushPopCompartment) {
    s.__next_compartment = next
}

func (s *StateVarPushPop) Increment() int {
    __e := StateVarPushPopFrameEvent{_message: "Increment"}
    __ctx := StateVarPushPopFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result int
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(int) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *StateVarPushPop) GetCount() int {
    __e := StateVarPushPopFrameEvent{_message: "GetCount"}
    __ctx := StateVarPushPopFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result int
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(int) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *StateVarPushPop) SaveAndGo() {
    __e := StateVarPushPopFrameEvent{_message: "SaveAndGo"}
    __ctx := StateVarPushPopFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *StateVarPushPop) Restore() {
    __e := StateVarPushPopFrameEvent{_message: "Restore"}
    __ctx := StateVarPushPopFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *StateVarPushPop) _state_Other(__e *StateVarPushPopFrameEvent) {
    __sv_comp := s.__compartment
    for __sv_comp != nil && __sv_comp.state != "Other" { __sv_comp = __sv_comp.parentCompartment }
    if __e._message == "$>" {
        if _, ok := __sv_comp.stateVars["other_count"]; !ok {
            __sv_comp.stateVars["other_count"] = 100
        }
    } else if __e._message == "GetCount" {
        s._context_stack[len(s._context_stack)-1]._return = __sv_comp.stateVars["other_count"].(int)
        return
    } else if __e._message == "Increment" {
        __sv_comp.stateVars["other_count"] = __sv_comp.stateVars["other_count"].(int) + 1
        s._context_stack[len(s._context_stack)-1]._return = __sv_comp.stateVars["other_count"].(int)
        return
    } else if __e._message == "Restore" {
        __popped := s._state_stack[len(s._state_stack)-1]
        s._state_stack = s._state_stack[:len(s._state_stack)-1]
        s.__transition(__popped)
        return
    }
}

func (s *StateVarPushPop) _state_Counter(__e *StateVarPushPopFrameEvent) {
    __sv_comp := s.__compartment
    for __sv_comp != nil && __sv_comp.state != "Counter" { __sv_comp = __sv_comp.parentCompartment }
    if __e._message == "$>" {
        if _, ok := __sv_comp.stateVars["count"]; !ok {
            __sv_comp.stateVars["count"] = 0
        }
    } else if __e._message == "GetCount" {
        s._context_stack[len(s._context_stack)-1]._return = __sv_comp.stateVars["count"].(int)
        return
    } else if __e._message == "Increment" {
        __sv_comp.stateVars["count"] = __sv_comp.stateVars["count"].(int) + 1
        s._context_stack[len(s._context_stack)-1]._return = __sv_comp.stateVars["count"].(int)
        return
    } else if __e._message == "SaveAndGo" {
        s._state_stack = append(s._state_stack, s.__compartment.copy())
        __compartment := newStateVarPushPopCompartment("Other")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    }
}

func main() {
	fmt.Println("=== Test 12: State Variable Push/Pop ===")
	sm := NewStateVarPushPop()

	// Increment counter to 3
	sm.Increment()
	sm.Increment()
	sm.Increment()
	count := sm.GetCount()
	if count != 3 {
		fmt.Printf("FAIL: Expected 3, got %d\n", count)
		os.Exit(1)
	}
	fmt.Printf("Counter before push: %d\n", count)

	// Push and go to Other state
	sm.SaveAndGo()
	fmt.Println("Pushed and transitioned to Other")

	// In Other state, count should be 100 (Other's state var)
	count = sm.GetCount()
	if count != 100 {
		fmt.Printf("FAIL: Expected 100 in Other state, got %d\n", count)
		os.Exit(1)
	}
	fmt.Printf("Other state count: %d\n", count)

	// Increment in Other
	sm.Increment()
	count = sm.GetCount()
	if count != 101 {
		fmt.Printf("FAIL: Expected 101 after increment, got %d\n", count)
		os.Exit(1)
	}
	fmt.Printf("Other state after increment: %d\n", count)

	// Pop back - should restore Counter with count=3
	sm.Restore()
	fmt.Println("Popped back to Counter")

	count = sm.GetCount()
	if count != 3 {
		fmt.Printf("FAIL: Expected 3 after pop (preserved), got %d\n", count)
		os.Exit(1)
	}
	fmt.Printf("Counter after pop: %d\n", count)

	// Increment to verify it works
	sm.Increment()
	count = sm.GetCount()
	if count != 4 {
		fmt.Printf("FAIL: Expected 4, got %d\n", count)
		os.Exit(1)
	}
	fmt.Printf("Counter after increment: %d\n", count)

	fmt.Println("PASS: State variables preserved across push/pop")
}

File: output/javascript/tests/12_state_var_push_pop.js

export class StateVarPushPopFrameEvent {
    _message;
    _parameters;

    constructor(message, parameters) {
        this._message = message;
        this._parameters = parameters;
    }
}


export class StateVarPushPopFrameContext {
    event;
    _return;
    _data;

    constructor(event, default_return) {
        this.event = event;
        this._return = default_return;
        this._data = {  };
    }
}


export class StateVarPushPopCompartment {
    state;
    state_args;
    state_vars;
    enter_args;
    exit_args;
    forward_event;
    parent_compartment;

    constructor(state, parent_compartment = null) {
        this.state = state;
        this.state_args = {  };
        this.state_vars = {  };
        this.enter_args = {  };
        this.exit_args = {  };
        this.forward_event = null;
        this.parent_compartment = parent_compartment;
    }

    copy() {
        const c = new StateVarPushPopCompartment(this.state, this.parent_compartment);
        c.state_args = {...this.state_args};
        c.state_vars = {...this.state_vars};
        c.enter_args = {...this.enter_args};
        c.exit_args = {...this.exit_args};
        c.forward_event = this.forward_event;
        return c;
    }
}


export class StateVarPushPop {
    _state_stack;
    __compartment;
    __next_compartment;
    _context_stack;

    constructor() {
        this._state_stack = [];
        this._context_stack = [];
        this.__compartment = new StateVarPushPopCompartment("Counter");
        this.__next_compartment = null;
        const __frame_event = new StateVarPushPopFrameEvent("$>", null);
        this.__kernel(__frame_event);
    }

    __kernel(__e) {
        // Route event to current state
        this.__router(__e);
        // Process any pending transition
        while (this.__next_compartment !== null) {
            const next_compartment = this.__next_compartment;
            this.__next_compartment = null;
            // Exit current state
            const exit_event = new StateVarPushPopFrameEvent("<$", this.__compartment.exit_args);
            this.__router(exit_event);
            // Switch to new compartment
            this.__compartment = next_compartment;
            // Enter new state (or forward event)
            if (next_compartment.forward_event === null) {
                const enter_event = new StateVarPushPopFrameEvent("$>", this.__compartment.enter_args);
                this.__router(enter_event);
            } else {
                // Forward event to new state
                const forward_event = next_compartment.forward_event;
                next_compartment.forward_event = null;
                if (forward_event._message === "$>") {
                    // Forwarding enter event - just send it
                    this.__router(forward_event);
                } else {
                    // Forwarding other event - send $> first, then forward
                    const enter_event = new StateVarPushPopFrameEvent("$>", this.__compartment.enter_args);
                    this.__router(enter_event);
                    this.__router(forward_event);
                }
            }
        }
    }

    __router(__e) {
        const state_name = this.__compartment.state;
        const handler_name = `_state_${state_name}`;
        const handler = this[handler_name];
        if (handler) {
            handler.call(this, __e);
        }
    }

    __transition(next_compartment) {
        this.__next_compartment = next_compartment;
    }

    increment() {
        const __e = new StateVarPushPopFrameEvent("increment", null);
        const __ctx = new StateVarPushPopFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()._return;
    }

    get_count() {
        const __e = new StateVarPushPopFrameEvent("get_count", null);
        const __ctx = new StateVarPushPopFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()._return;
    }

    save_and_go() {
        const __e = new StateVarPushPopFrameEvent("save_and_go", null);
        const __ctx = new StateVarPushPopFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    restore() {
        const __e = new StateVarPushPopFrameEvent("restore", null);
        const __ctx = new StateVarPushPopFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    _state_Other(__e) {
        // HSM: Navigate to this state's compartment for state var access
        let __sv_comp = this.__compartment;
        while (__sv_comp !== null && __sv_comp.state !== "Other") {
            __sv_comp = __sv_comp.parent_compartment;
        }
        if (__e._message === "$>") {
            if (!("other_count" in __sv_comp.state_vars)) {
                __sv_comp.state_vars["other_count"] = 100;
            }
        } else if (__e._message === "get_count") {
            this._context_stack[this._context_stack.length - 1]._return = __sv_comp.state_vars["other_count"];
            return;;
        } else if (__e._message === "increment") {
            __sv_comp.state_vars["other_count"] = __sv_comp.state_vars["other_count"] + 1;
            this._context_stack[this._context_stack.length - 1]._return = __sv_comp.state_vars["other_count"];
            return;;
        } else if (__e._message === "restore") {
            this.__transition(this._state_stack.pop());
            return;
        }
    }

    _state_Counter(__e) {
        // HSM: Navigate to this state's compartment for state var access
        let __sv_comp = this.__compartment;
        while (__sv_comp !== null && __sv_comp.state !== "Counter") {
            __sv_comp = __sv_comp.parent_compartment;
        }
        if (__e._message === "$>") {
            if (!("count" in __sv_comp.state_vars)) {
                __sv_comp.state_vars["count"] = 0;
            }
        } else if (__e._message === "get_count") {
            this._context_stack[this._context_stack.length - 1]._return = __sv_comp.state_vars["count"];
            return;;
        } else if (__e._message === "increment") {
            __sv_comp.state_vars["count"] = __sv_comp.state_vars["count"] + 1;
            this._context_stack[this._context_stack.length - 1]._return = __sv_comp.state_vars["count"];
            return;;
        } else if (__e._message === "save_and_go") {
            this._state_stack.push(this.__compartment.copy());
            const __compartment = new StateVarPushPopCompartment("Other", this.__compartment.copy());
            this.__transition(__compartment);
            return;
        }
    }
}

function main() {
    console.log("=== Test 12: State Variable Push/Pop ===");
    const s = new StateVarPushPop();

    // Increment counter to 3
    s.increment();
    s.increment();
    s.increment();
    let count = s.get_count();
    if (count !== 3) {
        throw new Error(`Expected 3, got ${count}`);
    }
    console.log(`Counter before push: ${count}`);

    // Push and go to Other state
    s.save_and_go();
    console.log("Pushed and transitioned to Other");

    // In Other state, count should be 100 (Other's state var)
    count = s.get_count();
    if (count !== 100) {
        throw new Error(`Expected 100 in Other state, got ${count}`);
    }
    console.log(`Other state count: ${count}`);

    // Increment in Other
    s.increment();
    count = s.get_count();
    if (count !== 101) {
        throw new Error(`Expected 101 after increment, got ${count}`);
    }
    console.log(`Other state after increment: ${count}`);

    // Pop back - should restore Counter with count=3
    s.restore();
    console.log("Popped back to Counter");

    count = s.get_count();
    if (count !== 3) {
        throw new Error(`Expected 3 after pop (preserved), got ${count}`);
    }
    console.log(`Counter after pop: ${count}`);

    // Increment to verify it works
    s.increment();
    count = s.get_count();
    if (count !== 4) {
        throw new Error(`Expected 4, got ${count}`);
    }
    console.log(`Counter after increment: ${count}`);

    console.log("PASS: State variables preserved across push/pop");
}

main();
